sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+_Returns_: A linearized version of the `id` returned by `get_local_range()`.
+
+[source,c++]
+----
+bool leader() const;
+----
+_Returns_: `true` for exactly one work-item in the opportunistic-group, if the
+calling work-item is the leader of the opportunistic-group, and `false` for all
+other work-items in the opportunistic-group. The leader of the opportunistic
+group is guaranteed to be the work-item for which `get_local_id()` returns 0.
+
+
+==== Usage Example
+
+The following example shows an atomic pointer being incremented.
+It is expected that all the work-items in the sub_group will increment the
+atomic value, but we opportunistically capture the groups of work-items as they
+arrive to this point in the control flow.
+
+[source, c++]
+----
+template <sycl::memory_order Order,
+          sycl::memory_scope Scope,
+          sycl::access::address_space AddressSpace>
+int atomic_aggregate_inc(sycl::sub_group sg, sycl::atomic_ref<int, Order, Scope, AddressSpace> ptr) {
+
+  // get the set of work-items that called this function simultaneously
+  auto active_group = sycl::ext::oneapi::experimental::get_opportunistic_group(sub_group);
+
+  // increment the atomic once on behalf of all active work-items
+  int count = active_group.get_local_linear_range();
+  int old_value;
+  if (active_group.leader()) {
+    old_value = ptr.fetch_add(count);
+  }
+
+  // return the value the individual work-item might have received if it had worked alone
+  auto base = sycl::group_broadcast(active_group, old_value);
+  auto idx = active_group.get_local_linear_id();
+  return base + idx;
+
+}
+----
+
+
+=== Masked Sub-groups
+
+A masked sub-group is a non-contiguous subset of a sub-group, representing an
+arbitrary user-defined subset of work-items. The members of a masked sub-group
+are described by a bitmask, where a 1 denotes membership of the group.
+
+The work-items within a masked sub-group retain information about the original
+sub-group, and many member functions of the `masked_sub_group` class reflect
+this. Developers are strongly recommended to use other user-constructed groups
+that match their use-case, both for improved performance and a simplified
+mental model.
+
+NOTE: Masked sub-groups exist primarily to support experimentation with
+arbitrary subsets of work-items within a sub-group, and to support the
+migration of algorithms already expressed via masks.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental {
+
+masked_sub_group get_masked_sub_group(sub_group sg, sub_group_mask mask);
+
+} // namespace ext::oneapi::experimental
+----
+
+_Preconditions_: All work-items in `sg` with a corresponding bit set in `mask`
+must encounter this function in converged control flow.
+
+_Returns_: A `masked_sub_group` consisting of the work-items in `sg` with a
+corresponding bit set in `mask`.
+
+
+==== Creation
+
+Masked sub-groups are created by calls to the `get_masked_sub_group()`
+function, which applies a bitmask to an existing sub-group.
```

Comment:
I'll ask the obvious question ... why do the member functions return information about the parent group rather than information about the work-items in the group that was created from the mask?  This group class is not symmetrical with the others, and it's not clear to me why that is.

Reply:
The CUDA [Warp Shuffle Functions](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-shuffle-functions) allow developers to identify subsets of a warp (i.e. sub-group) using a bitmask.  Providing something similar would simplify migration from CUDA and may be useful for tuning on devices using the CUDA backend.

The behavior of their `__shfl_sync` function is close to the behavior of `sycl::select_from_group`, but there's one key difference. With `__shfl_sync`, the index of the source lane is always in the range [0, 32), and is not modified by the mask parameter.  With `sycl::select_from_group`, the index of the source lane is in the range [0, `get_local_range()`).

Specifying that `masked_sub_group::get_local_range()` and `masked_sub_group::get_local_id()` have equivalent behavior to `sub_group::get_local_range()` and `sub_group::get_local_id()` means that the correct `select_from_group` behavior falls out.

In effect, this bridges the gap by providing two kinds of group:
1) Groups where the mental model is that some existing group has been modified to identify a subset of work-items.
(In the case of `masked_sub_group`, the `sub_group` has been masked.)
2) Groups where the mental model is that a new group has been created to identify a subset of work-items.
(This covers all the other group types.)

Another alternative (which only occurred to me while writing this) might be to define the `masked_sub_group` member functions in terms of group membership (e.g. `get_local_id()` would return the number of bits set prior to the current work-item), but add special wording for the `select_*` functions, like:

> ```c++
> template <typename T>
> T select_from_group(masked_sub_group g, T x, masked_sub_group::id_type remote_local_id);
> ```
> _Constraints_: Available only if `T` is a trivially copyable type.
> _Returns_: The value of `x` from the work-item with the local id specified by `remote_local_id`. The value of `remote_local_id` is interpreted relative to the parent sub-group, and does not correspond to a work-item's index in the mask. `remote_local_id` may identify a work-item outside of the group (i.e. with its corresponding bit set to 0 in the mask), but the value returned in this case is unspecified.

I think these solutions are pretty similar -- it's mainly a question of where we want to document the special behavior.  If we believe it only impacts `select_*` functions, is it cleaner to document the differences there?

Reply:
I think your proposed alternative would add this special wording:

> The value of `remote_local_id` is interpreted relative to the parent sub-group, and does not correspond to a work-item's index in the mask.

only in the case when it is called for `masked_sub_group` (and not for any other type of group).  Correct?

That seems even weirder to me.  Of the two options, I think it's better to have the weird definition of the `masked_sub_group` member functions than to have a weird definition of `select_from_group` for only one input type.

Reply:
That's correct.  If you think this is weirder, I'm happy to keep things as they are.